File: perceptron.cpp
#include<vector>
#include<iostream>
#include <random>
#include <math.h>





double quad_error(std::vector<std::vector<double>> &data, std::vector<double> &data_y, std::vector<double> &weights){
    double sum = 0;
    int samples = data.size();
    for (int i =0; i < weights.size(); i++){
        double output = 0;
        for (int j=0; j<data[i].size(); j++){
            output = output + (data[i][j]*weights[j]);
        }
        sum = sum + pow((output-data_y[i]),2.0);
    }
    return sum/samples;
}


void read_weights(std::vector<double> &weights) {
    for(int i=0; i<weights.size()-1;i++){
        std::cout << "weights vector is:" << weights[i] << std::endl;
    }
}










void percepetron(std::vector<std::vector<double>> &data, std::vector<double> &data_y){
    
    //pushing 1's on every single observation, to take in consideration the threshold weight.
    for(int i =0; i<data.size();i++){
        data[i].push_back(1);
    }
    

    //randomizing the first initialization of the weights vector.
    std::default_random_engine generator;
    std::uniform_real_distribution<double> distribution(0.0,1.0);

    /* initializing the weights vector for the first time.*/
    std::vector<double> weights;
    for (int i = 0; i < data[0].size()-1; i++)
        weights.push_back(distribution(generator));
    
    
    

    while(quad_error(data,data_y,weights)>1){
        
        std::cout <<"error:" << '\t' << quad_error(data,data_y,weights) << '\n';
        read_weights(weights);

        //calculating the perceptron output to find the error. 
        for (int i =0; i < data.size()-1; i++){
            double output = 0;
            for (int j=0; j<data[i].size()-1; j++){
                output = output + (data[i][j]*weights[j]);
            }
            //update the weights.
            double error = (data_y[i]- output);
            for(int k =0; k < weights.size()-1; k++){
                if (error<0){
                    weights[k] = weights[k] + data[i][k];
                } else {
                    weights[k] = weights[k] - data[i][k];
                }      
            }
        }
    }
}



int main() {
    std::vector<std::vector<double>> data = {
        {1,0,0},
        {0,3,0},
        {0,0,2}
    };

    std::vector<double> data_y = {1,3,2};

    percepetron(data,data_y);

    return 0;
}

File: perceptron.py
import random


y = [52,59,67,73,64,74,54,61,65,46,72]
data = [[132,137],[153,194],[143,184],[162,211],
[154,196],[168,220],[137,188],[149,188],[159,207],[128,167],[166,217]]




def learn(data,weights,learn_rate,iterations):
    k = 0
    while(True):
        for row in data:
            sumerror =0
            prediction = predict(row,weights)
            error = (prediction-row[-1])
            sumerror = sumerror + error**2
            if(sumerror<2):
                print("i think those are the best weights"+str(weights))
                break
            print("error:"+str(sumerror))
            for i in range(len(weights)):
                weights[i] = weights[i] - 0.00001*(error)*row[i]
        k+=1
        print(weights)


        



       



def predict(data,weights):
    output = 0
    for i in range(len(weights)):
        output = output + data[i]*weights[i]
    return output




          



            



def perceptron(data,output):
    #appending each observation data with 1,
    # to take in consideration the threshold weight.
    for i in data:
        i.append(1)
    for i in range(len(data)):
        data[i].append(output[i])
    dim = len(data[1])-1
    #initializing the weights vector.
    weights = []
    for i in range(0,dim):
        weights.append(0)
    
    learn(data,weights,0.0001,20)
    
    


    



    

perceptron(data,y)